includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-interface.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+interface Sensei_Reports_Overview_DataProvider_Interface {
+	/**
+	 * Get the data for the overview report.
+	 *
+	 * @param array $args
+	 * @param string|null $date_from
+	 * @param string|null $date_to
+	 *
+	 * @return array
+	 */
+	public function get_items( array $args, $date_from = null, $date_to = null ): array;
```

Comment:
Based on [this comment](https://github.com/Automattic/sensei/pull/4938#issuecomment-1075368945), maybe we could remove the`$date_from` and `$date_to` params and pass the dates with the `$args` or make a more universal `$filters` param and pass the dates there?

Reply:
I like your idea of passing them with `$args`. However, I think we should prepend names of these keys with something special (like that: `courses_date_from` and `courses_date_to`) to avoid confusion with query arguments.
And it would be better to rename `$args` to `$filters` then.
It is also better aligned with data providers' abstraction.

The other thing that I want to add is an abstract `get_date_filters` method to list table. It allows us to decide in concrete class what to pass there.

Reply:
> And it would be better to rename $args to $filters then.

`$filters` sound good to me. It's less confusing compared to `$args` which could be mixed up with query args. Another option would be to have both `$args` and `$filters` so they are not mixed up. Not sure if this suggestion makes it simpler or more complex. 😄 I guess it would make it simpler for someone to understand what we are doing and add a bit more complexity to the implementation side.

> However, I think we should prepend names of these keys with something special (like that: courses_date_from and courses_date_to or lessons_course_id) to avoid confusion with query arguments.

I think we should avoid having the `type` (courses, lessons) as a filter prefix because some filters are used on multiple types and the type doesn't give context about the filter itself. For example, we have the same date filters on both students and courses. My suggestion would be to use something like `last_activity_date_from` so it's a bit more specific to the data used to filter the items (last activity) instead of the type of items being filtered.

> The other thing that I want to add is an abstract get_date_filters get_additional_filters method to list table. It allows us to decide in concrete class what to pass there.

Yeah, makes sense! I'm not sure about the name. We have only one abstraction for filters so `get_filters` maybe? 🤔  It could be confused with WP filters, but I think people will figure out really quickly that it has nothing to do with that.

Reply:
> $filters sound good to me. It's less confusing compared to $args which could be mixed up with query args. Another option would be to have both $args and $filters so they are not mixed up. Not sure if this suggestion makes it simpler or more complex.

Yeah, I checked that these "args" differ from query args and, generally, they look like filters :) My initial desire was to introduce the second parameter `$filters`, but after a one more look at code I decided it might confuse people: why we have two kinds of filters? :) It might be hard for us to decide where to put one or another filtering parameter too.

> I think we should avoid having the type (courses, lessons) as a filter prefix because some filters are used on multiple types and the type doesn't give context about the filter itself.

Good point.

> I'm not sure about the name. We have only one abstraction for filters so get_filters maybe? 🤔

In current implementation we prepare part of filters in `generate_report` and `prepare_items` methods. These filters depend on the user will look at the report on the page or download it. (I've been considering extracting it to a separate method, but decided not to move it for now. But we can if we want :)) Also, another part of filters depends on our type (courses/students/lessons) of the report and we prepare it in a separate method. For me, this part looks like an additional one :) But if the name is not clear or still might confuse, let's rename it.
You can look at the usage here: 
- `prepare_items`: https://github.com/Automattic/sensei/pull/4938/files#diff-3b0db8ab59647710de5201ea0742e71848c10e6e1cd15321ef8d65fd3b736f05R115-R116 
- `generate_report`: https://github.com/Automattic/sensei/pull/4938/files#diff-3b0db8ab59647710de5201ea0742e71848c10e6e1cd15321ef8d65fd3b736f05R166-R167

Reply:
I'm ok with either approach and I think you did a great job with the structure! 🙇